src/client/dfuse/ops/read.c

```diff
@@ -102,6 +102,334 @@ dfuse_readahead_reply(fuse_req_t req, size_t len, off_t position, struct dfuse_o
 	return true;
 }
 
+static struct dfuse_eq *
+pick_eqt(struct dfuse_info *dfuse_info)
+{
+	uint64_t eqt_idx;
+
+	eqt_idx = atomic_fetch_add_relaxed(&dfuse_info->di_eqt_idx, 1);
+	return &dfuse_info->di_eqt[eqt_idx % dfuse_info->di_eq_count];
+}
+
+/* Chunk read and coalescing
+ *
+ * This code attempts to predict application and kernel I/O patterns and preemptively read file
+ * data ahead of when it's requested.
+ *
+ * For some kernels read I/O size is limited to 128k when using the page cache or 1Mb when using
+ * direct I/O.  To get around the performance impact of them detect when well aligned 128k reads
+ * are received and read an entire buffers worth, then for future requests the data should already
+ * be in cache.
```

Comment:
does the limitation of 128k alignment come from fuse?
i would not call only doing 128k reads from an application "well behaved". but im struggling to understand really when is this feature is actually used and what is the benefit.

If app is doing 4k reads, is the kernel reading in 128k blocks and this stores the chunks? or is this code used only if an app is doing 128k aligned IO?

Reply:
Inside chunk_read(),

	if (len != K128)
		return false;
	if ((position % K128) != 0)
		return false;

chunk_read() seems only handles special case of reading length ==128K and offset aligned on 128K.

Reply:
Yes.  For reads which go through the page cache then fuse will split the read into 128k chunks and submit them in parallel which is really what this feature is trying to address.

This is about ensuring that dfuse is delivering peak performance for well behaved or larger applications which I'd expect to be issuing I/Os in MB chunks - and in that case this deliveries significant benefit.

If the kernel sees 4k reads then dfuse will see 4k reads and this code won't be activated - the 4k read will be sent to the network.  I've written it in a way that could be extended to that use-case, it would need to use an extent map rather than the current `slots` and I question if an application reading 4k at a time would be as likely to end up reading the entire buffer so that would sit at a different place on the cost/benefit curve.  It would be good to have that as a option but it's probably better for the system not to enable it by default.